path/path.go

```diff
@@ -3,188 +3,301 @@ package path
 }
 
-// PopLastSegment returns a new Path without its final segment, and the final
-// segment, separately. If there is no more to pop (the path is just a key),
-// the original path is returned.
-func (p Path) PopLastSegment() (Path, string, error) {
+func (p *path) Segments() []string {
+	// Trim slashes from beginning and end, such that we do not return empty segments.
+	str := strings.TrimSuffix(p.str, "/")
+	str = strings.TrimPrefix(str, "/")
 
-	if p.IsJustAKey() {
-		return p, "", nil
+	return strings.Split(str, "/")
+}
+
+type resolvedPath struct {
+	path
+	cid       cid.Cid
+	remainder string
+}
+
+func (p *resolvedPath) Cid() cid.Cid {
+	return p.cid
+}
+
+func (p *resolvedPath) Remainder() string {
+	return p.remainder
+}
+
+// NewIPFSPath returns a new "/ipfs" path with the provided CID.
+func NewIPFSPath(cid cid.Cid) ResolvedPath {
+	return &resolvedPath{
+		path: path{
+			str:       fmt.Sprintf("/%s/%s", IPFSNamespace, cid.String()),
+			root:      cid,
+			namespace: IPFSNamespace,
+		},
+		cid:       cid,
+		remainder: "",
 	}
+}
 
-	segs := p.Segments()
-	newPath, err := ParsePath("/" + strings.Join(segs[:len(segs)-1], "/"))
-	if err != nil {
-		return "", "", err
+// NewIPLDPath returns a new "/ipld" path with the provided CID.
+func NewIPLDPath(cid cid.Cid) ResolvedPath {
+	return &resolvedPath{
+		path: path{
+			str:       fmt.Sprintf("/%s/%s", IPLDNamespace, cid.String()),
+			root:      cid,
+			namespace: IPLDNamespace,
+		},
+		cid:       cid,
+		remainder: "",
 	}
+}
 
-	return newPath, segs[len(segs)-1], nil
+// NewIPNSPath returns a new "/ipns" path with the provided CID.
+func NewIPNSPath(cid cid.Cid) Path {
+	return &path{
+		str:       fmt.Sprintf("/%s/%s", IPNSNamespace, cid.String()),
+		root:      cid,
+		namespace: IPNSNamespace,
+	}
 }
 
-// FromSegments returns a path given its different segments.
-func FromSegments(prefix string, seg ...string) (Path, error) {
-	return ParsePath(prefix + strings.Join(seg, "/"))
+func NewDNSLinkPath(domain string) Path {
+	return &path{
+		str:       fmt.Sprintf("/%s/%s", IPNSNamespace, domain),
+		root:      cid.Undef,
+		namespace: IPNSNamespace,
+	}
 }
 
-// ParsePath returns a well-formed ipfs Path.
-// The returned path will always be prefixed with /ipfs/ or /ipns/.
-// The prefix will be added if not present in the given string.
-// This function will return an error when the given string is
-// not a valid ipfs path.
-func ParsePath(txt string) (Path, error) {
-	parts := strings.Split(txt, "/")
-	if len(parts) == 1 {
-		kp, err := ParseCidToPath(txt)
+// NewPath returns a well-formed [Path]. The returned path will always be prefixed
+// with a valid namespace (/ipfs, /ipld, or /ipns). The prefix will be added if not
+// present in the given string. The rules are:
+//
+//  1. If the path has a single component (no slashes) ans it is a valid CID,
+//     an /ipfs path is returned. If the CID is encoded with the Libp2pKey codec,
+//     then a /ipns path is returned.
```

Comment:
Ok. So this is the behaviour of the current code in main. I am updating that as I agree this is a bit of a mess. This has been changed in [e688ad9](https://github.com/ipfs/boxo/pull/334/commits/e688ad9a343083239029d59652ea31c92b923063).  I had to make some updates in the namesys package to keep accepting legacy DNSLink records.

Reply:
I do agree with you. I think we should remove all the implied namespaces, including for IPFS. I think that will break some other things, at least within Boxo, but we can patch them where needed. From what I can see, it will only affect the processing of DNSLink entries.

Reply:
I pushed 4ba0583b76f365119ddc2e46c22a004a2a81192f, which removes all assumptions.  I do think that's the best way forward.